### ISSUE TITLE
test_remove_column_with_multi_column_index gets error with MariaDB 10.2.8

### ISSUE THREAD

**Issue by yahonda** [rails/rails#30485]
### Steps to reproduce

* Install MariaDB 10.2.8 
```ruby
$ ARCONN=mysql2 bin/test test/cases/migration/columns_test.rb:144
Using mysql2
Run options: --seed 32800

# Running:

E

Finished in 0.297244s, 3.3642 runs/s, 3.3642 assertions/s.

  1) Error:
ActiveRecord::Migration::ColumnsTest#test_remove_column_with_multi_column_index:
ActiveRecord::StatementInvalid: Mysql2::Error: Key column 'hat_size' doesn't exist in table: ALTER TABLE `test_models` DROP `hat_size`
    /var/lib/gems/2.4.0/gems/mysql2-0.4.9/lib/mysql2/client.rb:120:in `_query'
    /var/lib/gems/2.4.0/gems/mysql2-0.4.9/lib/mysql2/client.rb:120:in `block in query'
    /var/lib/gems/2.4.0/gems/mysql2-0.4.9/lib/mysql2/client.rb:119:in `handle_interrupt'
    /var/lib/gems/2.4.0/gems/mysql2-0.4.9/lib/mysql2/client.rb:119:in `query'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:187:in `block (2 levels) in execute'
    /home/ubuntu/rails/activesupport/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
    /home/ubuntu/rails/activesupport/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
    /home/ubuntu/rails/activesupport/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:186:in `block in execute'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:552:in `block (2 levels) in log'
    /usr/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:551:in `block in log'
    /home/ubuntu/rails/activesupport/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:542:in `log'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:185:in `execute'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb:28:in `execute'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:601:in `remove_column'
    /home/ubuntu/rails/activerecord/lib/active_record/connection_adapters/mysql/schema_statements.rb:54:in `remove_column'
    /home/ubuntu/rails/activerecord/test/cases/migration/helper.rb:38:in `remove_column'
    /home/ubuntu/rails/activerecord/test/cases/migration/columns_test.rb:150:in `test_remove_column_with_multi_column_index'

1 runs, 1 assertions, 0 failures, 1 errors, 0 skips
$
```

### Expected behavior
Tell us what should happen

### Actual behavior
Tell us what happens instead

### System configuration
**Rails version**:master branch

**Ruby version**:ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux-gnu]

** MariaDB version**; Server version: 10.2.8-MariaDB-10.2.8+maria~zesty-log mariadb.org binary distribution

### Other information

Here is the debug.log for `test_models` table related operations
* debug.log
```ruby
   (40.5ms)  CREATE TABLE `test_models` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `created_at` datetime, `updated_at` datetime) ENGINE=InnoDB
   (77.9ms)  ALTER TABLE `test_models` ADD `hat_size` int
   (86.6ms)  ALTER TABLE `test_models` ADD `hat_style` varchar(100)
   (46.1ms)  CREATE UNIQUE INDEX `index_test_models_on_hat_style_and_hat_size`  ON `test_models` (`hat_style`, `hat_size`)
   (0.4ms)  ALTER TABLE `test_models` DROP `hat_size`
   (20.7ms)  DROP TABLE IF EXISTS `test_models`
```

* I am also able to reproduce this without ActiveRecord
```sql
MariaDB [(none)]> use test;
Database changed
MariaDB [test]> CREATE TABLE `test_models` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `created_at` datetime, `updated_at` datetime) ENGINE=InnoDB;
Query OK, 0 rows affected (0.05 sec)

MariaDB [test]> desc test_models;
+------------+------------+------+-----+---------+----------------+
| Field      | Type       | Null | Key | Default | Extra          |
+------------+------------+------+-----+---------+----------------+
| id         | bigint(20) | NO   | PRI | NULL    | auto_increment |
| created_at | datetime   | YES  |     | NULL    |                |
| updated_at | datetime   | YES  |     | NULL    |                |
+------------+------------+------+-----+---------+----------------+
3 rows in set (0.00 sec)

MariaDB [test]> ALTER TABLE `test_models` ADD `hat_size` int;
Query OK, 0 rows affected (0.08 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [test]> desc test_models;
+------------+------------+------+-----+---------+----------------+
| Field      | Type       | Null | Key | Default | Extra          |
+------------+------------+------+-----+---------+----------------+
| id         | bigint(20) | NO   | PRI | NULL    | auto_increment |
| created_at | datetime   | YES  |     | NULL    |                |
| updated_at | datetime   | YES  |     | NULL    |                |
| hat_size   | int(11)    | YES  |     | NULL    |                |
+------------+------------+------+-----+---------+----------------+
4 rows in set (0.00 sec)

MariaDB [test]> ALTER TABLE `test_models` ADD `hat_style` varchar(100);
Query OK, 0 rows affected (0.09 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [test]>  desc test_models;
+------------+--------------+------+-----+---------+----------------+
| Field      | Type         | Null | Key | Default | Extra          |
+------------+--------------+------+-----+---------+----------------+
| id         | bigint(20)   | NO   | PRI | NULL    | auto_increment |
| created_at | datetime     | YES  |     | NULL    |                |
| updated_at | datetime     | YES  |     | NULL    |                |
| hat_size   | int(11)      | YES  |     | NULL    |                |
| hat_style  | varchar(100) | YES  |     | NULL    |                |
+------------+--------------+------+-----+---------+----------------+
5 rows in set (0.00 sec)

MariaDB [test]> CREATE UNIQUE INDEX `index_test_models_on_hat_style_and_hat_size`  ON `test_models` (`hat_style`, `hat_size`);
Query OK, 0 rows affected (0.04 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [test]> ALTER TABLE `test_models` DROP `hat_size`;
ERROR 1072 (42000): Key column 'hat_size' doesn't exist in table
MariaDB [test]> DROP TABLE IF EXISTS `test_models`
    -> ;
Query OK, 0 rows affected (0.03 sec)

MariaDB [test]>
```

* If `index_test_models_on_hat_style_and_hat_size` index is not created, drop column work fine.

```sql
MariaDB [(none)]> use test;
Database changed
MariaDB [test]> CREATE TABLE `test_models` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `created_at` datetime, `updated_at` datetime) ENGINE=InnoDB;
Query OK, 0 rows affected (0.04 sec)

MariaDB [test]> ALTER TABLE `test_models` ADD `hat_size` int;
Query OK, 0 rows affected (0.08 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [test]> ALTER TABLE `test_models` ADD `hat_style` varchar(100);
Query OK, 0 rows affected (0.07 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [test]> ALTER TABLE `test_models` DROP `hat_size`;
Query OK, 0 rows affected (0.06 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [test]> DROP TABLE IF EXISTS `test_models`;
Query OK, 0 rows affected (0.01 sec)

MariaDB [test]>
```

**Comment by yahonda** [comment]
This behavior is explained at [DROP COLUMN [IF EXISTS] col_name [CASCADE|RESTRICT]](https://mariadb.com/kb/en/the-mariadb-library/alter-table/#drop-column-if-exists-col_name-cascaderestrict)

This is an expected behavior since MariaDB 10.2.8.

**Comment by yahonda** [comment]
I have pushed https://github.com/rails/rails/pull/29850/commits/6add4c74d628c69a15cf611743b089fa1f696220 to #29850 since without this commit CI never goes green with MariaDB 10.2

**Comment by yahonda** [comment]
Addressed by #29850
